invenio_indexer/cli.py

```diff
@@ -28,27 +28,55 @@
 
 import click
 from flask_cli import with_appcontext
+from invenio_records.models import RecordMetadata
 from invenio_search.cli import index
 
 from .api import RecordIndexer
 from .tasks import process_bulk_queue
 
 
+def abort_if_false(ctx, param, value):
+    """Abort command is value is False."""
+    if not value:
+        ctx.abort()
+
+
 @index.command()
 @click.option(
-    '--delayed', '-d', is_flag=True, help="Run indexing in background.")
+    '--delayed', '-d', is_flag=True, help='Run indexing in background.')
 @click.option(
     '--concurrency', '-c', default=1, type=int,
-    help="Number of concurrent indexing tasks to start.")
+    help='Number of concurrent indexing tasks to start.')
 @with_appcontext
 def run(delayed, concurrency):
     """Run bulk record indexing."""
     if delayed:
         click.secho(
-            "Starting {0} tasks for indexing records...".format(concurrency),
+            'Starting {0} tasks for indexing records...'.format(concurrency),
             fg='green')
         for c in range(0, concurrency):
             process_bulk_queue.delay()
     else:
-        click.secho("Indexing records...", fg='green')
+        click.secho('Indexing records...', fg='green')
         RecordIndexer().process_bulk_queue()
+
+
+@index.command()
+@click.option('--yes-i-know', is_flag=True, callback=abort_if_false,
+              expose_value=False,
+              prompt='Do you really want to reindex all records?')
+@with_appcontext
+def reindex():
+    """Reindex all records.
+
+    NOTE: Deleted records are not removed from the index.
+    """
+    click.secho('Sending records to indexing queue ...', fg='green')
+
+    def records():
+        """Record iterator."""
+        for record in RecordMetadata.query.values(RecordMetadata.id):
+            yield record[0]
+
+    RecordIndexer().bulk_index(records())
```

Comment:
I prefer to keep explicit iterator over building full list of UUIDs. I could replace it with:

```python
RecordIndexer().bulk_index(r[0] for r in RecordMetadata.query.values(RecordMetadata.id))
```

Then I should break long line hence:

```python
RecordIndexer().bulk_index(
    r[0] for r in RecordMetadata.query.values(RecordMetadata.id)
)
```

The truth is I will save some lines, but later we want to introduce some filtering, hence the `RecordMetadata.query` part will have to be extended and reading from arguments or `stdin` will need to be added.

At the end, I can rewrite it to

```python
records  = (r[0] for r in RecordMetadata.query.values(RecordMetadata.id))
RecordIndexer().bulk_index(records)
```

Does it make it more readable? Maybe yes. I should definitely add a comment so nobody refactors the iterator to a list comprehension.

```python
# NOTE records should be an iterator to save memory.
records  = (r[0] for r in RecordMetadata.query.values(RecordMetadata.id))
RecordIndexer().bulk_index(records)
```